src/chat.cpp

```diff
@@ -313,22 +340,115 @@ u32 ChatBuffer::formatChatLine(const ChatLine& line, u32 cols,
+			else
 			{
-				if (iswspace(line.text.getString()[in_pos + frag_length]))
-					space_pos = frag_length;
-				++frag_length;
+				u32 http_pos = u32(std::wstring::npos);
+				text_processing = false;  // set FALSE at end of this block
+				bool halt_fragloop = false;
+				while(!halt_fragloop)
+				{
+					remaining_in_input = line.text.size() - in_pos;
+					frag_length = 0;
+					space_pos = 0;
+
+					// Note: unsigned(-1) on fail
+					http_pos = u32(line.text.getString().find(L"https://", std::size_t(in_pos)));
+					if(http_pos == u32(std::wstring::npos))
+						http_pos = u32(line.text.getString().find(L"http://", std::size_t(in_pos)));
+					if(http_pos != u32(std::wstring::npos))
+						http_pos -= in_pos;
+
+					while (frag_length < remaining_in_input &&
+							frag_length < remaining_in_output)
+					{
+						if (iswspace(line.text.getString()[in_pos + frag_length]))
+							space_pos = frag_length;
+						++frag_length;
+					}
+
+					// Http not in range, grab until space or EOL, halt as normal.
+					if(http_pos > remaining_in_output)  // sufficient unless screen is infinitely wide
+					{
+						halt_fragloop = true;
+						// force text processing on THIS frag
+						text_processing = true;
+					}
+					// At http, grab ALL until FIRST whitespace or end marker. loop.
+					// If at end of string, next loop will be empty string to mark end of weblink. This is intentional.
+					else if(http_pos == 0)
+					{
+						frag_length = 6;
+						// TODO? replace this with a safer (slower) regex whitelist?
+						wchar_t tempchar = line.text.getString()[in_pos+frag_length];
+						while (frag_length < remaining_in_input &&
+								!iswspace(tempchar) &&
+								tempchar != L'\'' &&
+								tempchar != L'\"' &&
+								tempchar != L')' &&
+								tempchar != L';'
+							)
+						{
+							++frag_length;
+							tempchar = line.text.getString()[in_pos+frag_length];
+						}
+						space_pos = frag_length - 1;
+						if(frag_length >= remaining_in_output)
+						{
+							// Force text processing on THIS frag
+							text_processing = true;
+							halt_fragloop = true;
+						}
+					}
+					// Http in range, grab until http, loop
+					else
+					{
+						space_pos = http_pos - 1;
+						frag_length = http_pos;
+					}
+
+					if (space_pos != 0 && frag_length < remaining_in_input)
+						frag_length = space_pos + 1;
+
+					temp_frag.text = line.text.substr(in_pos, frag_length);
+					temp_frag.column = text_processing ? u32(INT_MAX) : 0;
+					if(http_pos == 0)
+					{
+						temp_frag.meta = wide_to_utf8(temp_frag.text.getString());
+						temp_frag.text = EnrichedString(temp_frag.text.getString());
+						temp_frag.text.setDefaultColor(m_cache_chat_weblink_color);
+					}
+					else
+					{
+						temp_frag.meta = "";
+					}
+					next_frags.push_back(temp_frag);
+					in_pos += frag_length;
+					remaining_in_output -= std::min(frag_length, remaining_in_output);
```

Comment:
It would be great to have some comments to explain the overall idea of what's supposed to happen here.

Please also apply the Minetest code style rules in those changes.